File: javaAbstractor/src/main/java/abstractor/core/constructs/Abstract.java
package abstractor.core.constructs;

import abstractor.core.cmp.Cmp;
import abstractor.core.cmp.CmpOptions;
import abstractor.core.json.*;

public class Abstract extends ConstructImp {
    public String         name;
    public Ref<Signature> signature;

    public Abstract() {}

    public Abstract(String name, Ref<Signature> signature) {
        this.name      = name;
        this.signature = signature;
    }

    public ConstructKind kind() { return ConstructKind.ABSTRACT; }

    @Override
    public JsonNode toJson(JsonHelper h) {
        JsonObject obj = (JsonObject)super.toJson(h);
        obj.put("name",      this.name);
        obj.put("signature", index(this.signature));
        return obj;
    }

    @Override
    public Cmp getCmp(Construct c, CmpOptions options) {
        return Cmp.or(super.getCmp(c, options),
            Cmp.defer(this.name,      () -> ((Abstract)c).name),
            Cmp.defer(this.signature, () -> ((Abstract)c).signature)
        );
    }   
}

File: goAbstractor2/internal/constructs/abstract.go
package constructs

import "go/types"

type Abstract struct {
	Name      string     `json:"name"`
	Signature *Signature `json:"signature"`
}

type AbstractFactory struct {
	*Factory[*types.Func, Abstract]
}

func newAbstractFactory(proj *Project) *AbstractFactory {
	return &AbstractFactory{
		Factory: NewFactory(proj, func(proj *Project, src *types.Func, ab *Abstract) {
			ab.Name = src.Name()
			ab.Signature = proj.Signatures.New(src.Signature())
		}),
	}
}

func (a Abstract) Kind() string { return `abstract` }
